File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> ans;
    vector<string> dial = {"0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    
    vector<string> letterCombinations(string digits) {
        ans.clear();
        if (digits.length() == 0) {
            return ans;
        }
        helper("", 0, digits);
        return ans;
    }
    
    void helper(string comb, int index, string digits) {
        if (index == digits.length()) {
            ans.push_back(comb);
            return;
        }
            
        string letters = dial[digits[index] - '0'];
        for (int i = 0; i < letters.length(); i++) {
            helper(comb + letters[i], index + 1, digits);
        }
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        if (digits.isEmpty()) return Collections.emptyList();
        String[] mapping = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs" ,"tuv", "wxyz"};
        List<String>l=new ArrayList<String>();
        solve(l,new String(),digits,mapping,0);
        return l;
    }
    public void solve(List<String>l,String s,String digits,String[] mapping,int i){
        if(i>=digits.length()){
            l.add(s);
            return;
        }
        int num=digits.charAt(i)-'0';
        String value=mapping[num];
        for(int j=0;j<value.length();j++){
            s+=value.charAt(j);
            solve(l,s,digits,mapping,i+1);
            s = s.substring(0, s.length() - 1);
        }
    }
}